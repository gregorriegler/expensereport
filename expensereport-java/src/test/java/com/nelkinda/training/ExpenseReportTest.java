package com.nelkinda.training;

import org.approvaltests.ApprovalUtilities;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;

import java.io.ByteArrayOutputStream;
import java.util.Arrays;
import java.util.Collections;
import java.util.Date;

public class ExpenseReportTest {

    private final ByteArrayOutputStream outputStream = new ApprovalUtilities().writeSystemOutToStringBuffer();
    private final ExpenseReport expenseReport = new ExpenseReport();

    @Test
    public void printEmptyReport() {
        expenseReport.printReport(Collections.emptyList(), new Date(0));

        Approvals.verify(outputStream.toString());
    }

    @Test
    public void printDinner() {
        expenseReport.printReport(Collections.singletonList(createExpense(ExpenseType.DINNER, 10)), new Date(0));

        Approvals.verify(outputStream.toString());
    }

    @Test
    public void printBreakfast() {
        expenseReport.printReport(Collections.singletonList(createExpense(ExpenseType.BREAKFAST, 10)), new Date(0));

        Approvals.verify(outputStream.toString());
    }

    @Test
    public void printCarRental() {
        expenseReport.printReport(Collections.singletonList(createExpense(ExpenseType.CAR_RENTAL, 10)), new Date(0));

        Approvals.verify(outputStream.toString());
    }

    @Test
    public void printDinnerOverExpense() {
        expenseReport.printReport(Collections.singletonList(createExpense(ExpenseType.DINNER, 5100)), new Date(0));

        Approvals.verify(outputStream.toString());
    }

    @Test
    public void printBreakfastOverExpense() {
        expenseReport.printReport(Collections.singletonList(createExpense(ExpenseType.BREAKFAST, 2000)), new Date(0));

        Approvals.verify(outputStream.toString());
    }

    @Test
    public void printDinnerAndBreakfast() {
        expenseReport.printReport(Arrays.asList(createExpense(ExpenseType.DINNER, 5100),
                createExpense(ExpenseType.BREAKFAST, 20)),
                new Date(0));

        Approvals.verify(outputStream.toString());
    }

    @Test
    public void printDinnerAndBreakfastAndCarRental() {
        expenseReport.printReport(Arrays.asList(createExpense(ExpenseType.DINNER, 5100),
                createExpense(ExpenseType.BREAKFAST, 20),
                createExpense(ExpenseType.DINNER, 50),
                createExpense(ExpenseType.CAR_RENTAL, 33234))
                , new Date(0));

        Approvals.verify(outputStream.toString());
    }

    private Expense createExpense(ExpenseType expenseType, int amount) {
        Expense expense = new Expense();
        expense.type = expenseType;
        expense.amount = amount;
        return expense;
    }

}
