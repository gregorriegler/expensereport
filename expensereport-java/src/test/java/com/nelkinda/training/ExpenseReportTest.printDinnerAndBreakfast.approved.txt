Expenses Thu Jan 01 01:00:00 CET 1970
Dinner	5100	X
Breakfast	20	 
Meal expenses: 5120
Total expenses: 5120
